File: android/src/main/kotlin/com/appdynamics/appdynamics_mobilesdk/features/user_data/UserDataLong.kt
/*
 * Copyright (c) 2021. AppDynamics LLC and its affiliates.
 * All rights reserved.
 *
 */

package com.appdynamics.appdynamics_mobilesdk.features

import androidx.annotation.NonNull
import com.appdynamics.appdynamics_mobilesdk.AppDynamicsMobileSdkPlugin
import com.appdynamics.eumagent.runtime.Instrumentation
import io.flutter.plugin.common.MethodChannel
import java.util.*

fun AppDynamicsMobileSdkPlugin.setUserDataLong(
    @NonNull result: MethodChannel.Result,
    arguments: Any?
) {
    val properties = arguments as HashMap<*, *>

    val key = properties["key"] as? String ?: run {
        result.error(
            "500",
            "Agent setUserDataLong() failed.",
            "Please provide a valid string for `key`."
        )
        return
    }

    val value = properties["value"].toString().toLong() ?: run {
        result.error(
            "500",
            "Agent setUserDataLong() failed.",
            "Please provide a valid long for `value`."
        )
        return
    }

    Instrumentation.setUserDataLong(key, value);
    result.success(null)
}


fun AppDynamicsMobileSdkPlugin.removeUserDataLong(
    @NonNull result: MethodChannel.Result,
    arguments: Any?
) {
    val key = arguments as? String ?: run {
        result.error(
            "500",
            "Agent removeUserDataLong() failed.",
            "Please provide a valid string for `key`."
        )
        return
    }

    Instrumentation.setUserDataLong(key, null);
    result.success(null)
}

File: ios/Classes/Features/User Data/UserDataLong.swift
import ADEUMInstrumentation
import Flutter

extension SwiftAppDynamicsAgentPlugin {
  func setUserDataLong(result: @escaping FlutterResult, arguments: Any?) {
    guard let properties = arguments as? Dictionary<String, Any> else {
      return
    }
    
    guard let key = properties["key"] as? String else {
      let error = FlutterError(code: "500", message: "Agent setUserDataLong() failed.", details: "Please provide a valid string for `key`.")
      result(error)
      return
    }
    
    guard let value = properties["value"] as? Int64 else {
      let error = FlutterError(code: "500", message: "Agent setUserDataLong() failed.", details: "Please provide a valid long for `value`.")
      result(error)
      return
    }
    
    ADEumInstrumentation.setUserDataLong(key, value: value)
    result(nil)
  }
  
  func removeUserDataLong(result: @escaping FlutterResult, arguments: Any?) {
    guard let key = arguments as? String else {
      let error = FlutterError(code: "500", message: "Agent removeUserDataLong() failed.", details: "Please provide a valid string for `key`.")
      result(error)
      return
    }
    
    ADEumInstrumentation.removeUserDataLong(key)
    result(nil)
  }
}
